File: python3/sort_the_matrix_diagonally.py
from collections import defaultdict

"""
Space   : O(nm)
Time    : O(nm log n)
"""


class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        m, n = len(mat), len(mat[0])
        d = defaultdict(list)

        for i in range(m):
            for j in range(n):
                d[i-j].append(mat[i][j])

        for k, _ in d.items():
            d[k].sort()

        for i in range(m):
            for j in range(n):
                mat[i][j] = d[i-j].pop(0)

        return mat

File: golang/sort_the_matrix_diagonally.go
package main

import "sort"

/*
Space	: O(mn)
Time	: O(mn* log n)
*/

func diagonalSort(mat [][]int) [][]int {
	m, n := len(mat), len(mat[0])
	d := make(map[int][]int)

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if _, ok := d[i-j]; ok {
				d[i-j] = append(d[i-j], mat[i][j])
			} else {
				d[i-j] = []int{mat[i][j]}
			}
		}
	}

	for k, _ := range d {
		sort.Ints(d[k])
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			num := d[i-j][0]
			d[i-j] = d[i-j][1:]
			mat[i][j] = num
		}
	}

	return mat
}
